lib/Containers/MerkleTree.cpp

```diff
@@ -166,17 +264,85 @@ MerkleTree<Hasher, BranchingBits>::fromSnappyCompressed(std::string_view buffer)
   }
   
   return std::unique_ptr<MerkleTree<Hasher, BranchingBits>>(
-      new MerkleTree<Hasher, BranchingBits>(std::move(uncompressed)));
+      new MerkleTree<Hasher, BranchingBits>(std::string_view(reinterpret_cast<char const*>(uncompressed.get()), length)));
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
 std::unique_ptr<MerkleTree<Hasher, BranchingBits>>
-MerkleTree<Hasher, BranchingBits>::fromBottomMostCompressed(std::string_view buffer) {
+MerkleTree<Hasher, BranchingBits>::fromSnappyLazyCompressed(std::string_view buffer) {
   char const* p = buffer.data();
   char const* e = p + buffer.size();
 
-  if (p + sizeof(Meta) > e) {
-    throw std::invalid_argument("invalid compressed tree data");
+  if (p + MetaSize - sizeof(Meta::Padding) + sizeof(uint32_t) >= e) {
+    throw std::invalid_argument("invalid compressed tree data in buffer (lazy compression)");
+  }
+  
+  Data data;
+
+  data.meta.rangeMin = readUInt<uint64_t>(p);
+  data.meta.rangeMax = readUInt<uint64_t>(p);
+  data.meta.depth = readUInt<uint64_t>(p);
+  data.meta.initialRangeMin = readUInt<uint64_t>(p);
+
+  std::uint64_t summaryCount = readUInt<uint64_t>(p);
+  std::uint64_t summaryHash = readUInt<uint64_t>(p);
+  data.meta.summary = { summaryCount, summaryHash };
+
+  TRI_ASSERT(p - buffer.data() == MetaSize - sizeof(Meta::Padding));
+ 
+  std::uint32_t numberOfShards = readUInt<uint32_t>(p);
+  if (numberOfShards >= 1024) {
+    throw std::invalid_argument("invalid compressed shard count in tree data");
+  }
+  // initialize all shards to nullptrs
+  data.shards.resize(numberOfShards);
+
+  char const* compressedData = p + numberOfShards * sizeof(uint32_t);
+  for (std::uint64_t i = 0; i < numberOfShards; ++i) {
+    TRI_ASSERT(p < e);
+    std::uint32_t compressedLength = readUInt<uint32_t>(p);
+    if (compressedLength != 0) {
+      size_t shardLength = shardSize(data.meta.depth);
+      TRI_ASSERT(shardLength % NodeSize == 0);
+      std::uint64_t nodesPerShard = shardLength / NodeSize;
+      // ShardSize may be more than we actually need (if we only have a single
+      // shard), but it doesn't matter for correctness here.
+      data.shards[i] = std::make_unique<uint8_t[]>(shardLength);
+      {
+        Node* ptr = reinterpret_cast<Node*>(data.shards[i].get());
+        for (std::uint64_t i = 0; i < nodesPerShard; ++i) {
+          new (ptr) Node{0, 0};
+          ++ptr;
+        }
+      }
+
+      size_t length;
+      bool canUncompress = snappy::GetUncompressedLength(compressedData, compressedLength, &length);
+      if (!canUncompress || length != shardLength) {
+        throw std::invalid_argument(
+            "Cannot determine size of Snappy-compressed data.");
+      }
+      if (!snappy::RawUncompress(compressedData, compressedLength, reinterpret_cast<char*>(data.shards[i].get()))) {
```

Comment:
I converted `struct Node` into a POD, so it should be compatible with memory operations on the byte level. I also moved the code away from placement new and instead used plain memset to initialize the memory.

Reply:
Although this should work as expected, IMO this is still not standard conform. We are allocating a memory block of type `uint8_t[]` but later access it as `Node[]`, and on top of it we cast it to `char*` in for the snappy uncompress.
I'm not saying we should change this code now, but we should be careful with code like this because AFAIK this is unspecified if not undefined behavior.

Reply:
FWIW: if we would allocate a `Node[]` instead of a `uint8_t[]` this would already improve the situation. As it is now it is not guaranteed that the allocation has the proper alignment required by `Node`.